### PR TITLE
Improve types for some array internals

### PR DESCRIPTION
**Pull request by wagenet**
It's hard to tell in this diff, but some of the EmberArray types changed a little bit to be more accurate.

### CODE REVIEW

File: packages/@ember/-internals/metal/lib/array.ts [hunk 1]
```diff
@@ -17,13 +19,14 @@ export function objectAt<T>(array: T[] | EmberArray<T>, index: number): T | unde
 }
 
 export function replace<T>(
-  array: NativeArray<T> | MutableArray<T>,
+  array: T[] | MutableArray<T>,
```
Comment (wagenet) [inline packages/@ember/-internals/metal/lib/array.ts:29]: This actually should handle a true native array vs an Ember.NativeArray.

File: packages/@ember/-internals/metal/lib/array.ts [hunk 2]
```diff
@@ -16,16 +19,23 @@ export function objectAt<T>(array: T[] | EmberArray<T>, index: number): T | unde
...
   start: number,
   deleteCount: number,
   items: T[] = EMPTY_ARRAY as []
 ): void {
-  if (Array.isArray(array)) {
-    replaceInNativeArray(array, start, deleteCount, items);
-  } else {
+  if (isMutableArray(array)) {
```
Comment (wagenet) [inline packages/@ember/-internals/metal/lib/array.ts:34]: `Array.isArray` will match as true for `Ember.NativeArray` but since that has a special `replace` method we want to use that instead.

File: packages/@ember/-internals/metal/lib/array.ts [hunk 3]
```diff
@@ -16,16 +19,23 @@ export function objectAt<T>(array: T[] | EmberArray<T>, index: number): T | unde
...
     array.replace(start, deleteCount, items);
+  } else {
+    assert('Can only replace content of a native array or MutableArray', Array.isArray(array));
```
Comment (wagenet) [inline packages/@ember/-internals/metal/lib/array.ts:37]: This is mainly just to keep TS happy.

File: packages/@ember/-internals/metal/lib/array.ts [hunk 4]
```diff
@@ -34,7 +44,7 @@ const CHUNK_SIZE = 60000;
 // To avoid overflowing the stack, we splice up to CHUNK_SIZE items at a time.
 // See https://code.google.com/p/chromium/issues/detail?id=56588 for more details.
 export function replaceInNativeArray<T>(
-  array: T[] | NativeArray<T>,
+  array: T[],
```
Comment (wagenet) [inline packages/@ember/-internals/metal/lib/array.ts:47]: `NativeArray<T>` is already `T[]`.

File: packages/@ember/-internals/runtime/lib/mixins/array.d.ts [hunk 5]
```diff
@@ -107,4 +107,4 @@ export function uniqBy<T>(
   array: T[] | EmberArray<T>,
   keyOrFunc: string | ((item: T) => unknown)
 ): T[] | EmberArray<T>;
-export function isArray(obj: unknown): boolean;
+export function isArray(obj: unknown): obj is ArrayLike<unknown> | EmberArray<unknown>;
```
Comment (wagenet) [inline packages/@ember/-internals/runtime/lib/mixins/array.d.ts:110]: This is pretty much what is actually checked. However, the actual code is kind of bonkers and we should probably stop using it!

File: packages/@ember/-internals/runtime/lib/system/array_proxy.ts [hunk 6]
```diff
@@ -30,7 +30,7 @@ import {
 } from '@glimmer/validator';
 import { PropertyDidChange } from '@ember/-internals/metal/lib/property_events';
 
-function isMutable<T>(obj: EmberArray<T>): obj is MutableArray<T> {
+function isMutable<T>(obj: T[] | EmberArray<T>): obj is T[] | MutableArray<T> {
```
Comment (wagenet) [inline packages/@ember/-internals/runtime/lib/system/array_proxy.ts:33]: The code below allows for a native array.

File: packages/@ember/-internals/runtime/lib/system/array_proxy.ts [hunk 7]
```diff
@@ -116,8 +116,8 @@ function customTagForArrayProxy(proxy: object, key: string) {
   @public
 */
 // eslint-disable-next-line @typescript-eslint/no-empty-interface, @typescript-eslint/no-unused-vars
-interface ArrayProxy<T, C extends EmberArray<T> = EmberArray<T>> extends MutableArray<T> {}
-class ArrayProxy<T, C extends EmberArray<T> = EmberArray<T>>
+interface ArrayProxy<T, C extends EmberArray<T> | T[] = T[]> extends MutableArray<T> {}
+class ArrayProxy<T, C extends EmberArray<T> | T[] = T[]>
```
Comment (wagenet) [inline packages/@ember/-internals/runtime/lib/system/array_proxy.ts:120]: We should be able to proxy to native arrays too.

File: packages/@ember/-internals/runtime/lib/system/array_proxy.ts [hunk 8]
```diff
@@ -218,15 +218,15 @@ class ArrayProxy<T, C extends EmberArray<T> = EmberArray<T>>
     @method replaceContent
     @param {Number} idx The starting index
     @param {Number} amt The number of items to remove from the content.
-    @param {EmberArray} objects Optional array of objects to insert.
+    @param {Array} objects Optional array of objects to insert.
```
Comment (wagenet) [inline packages/@ember/-internals/runtime/lib/system/array_proxy.ts:221]: This can't actually accept an EmberArray.

File: packages/@ember/-internals/runtime/lib/system/array_proxy.ts [hunk 9]
```diff
@@ -218,15 +218,15 @@ class ArrayProxy<T, C extends EmberArray<T> = EmberArray<T>>
...
     @return {void}
     @public
   */
   replaceContent(idx: number, amt: number, objects?: T[]) {
     let content = get(this, 'content');
-    assert('[BUG] Called objectAtContent without content', content);
+    assert('[BUG] Called replaceContent without content', content);
     assert('Mutating a non-mutable array is not allowed', isMutable(content));
-    content.replace(idx, amt, objects);
+    replace<T>(content, idx, amt, objects);
```
Comment (wagenet) [inline packages/@ember/-internals/runtime/lib/system/array_proxy.ts:229]: Allow us to replace a true native array.

File: packages/@ember/-internals/runtime/lib/mixins/array.ts [hunk 10]
```diff
@@ -209,7 +251,68 @@ function mapBy(key) {
...
+  mapBy<K extends string>(key: K): NativeArray<Value<T, K>>;
+  filter<Target>(
+    callback: (this: Target, item: T, index: number, arr: this) => unknown,
+    target?: Target
+  ): NativeArray<T>;
+  reject<Target>(
+    callback: (this: Target, item: T, index: number, arr: this) => unknown,
+    target?: Target
+  ): NativeArray<T>;
+  filterBy(key: string, value?: unknown): NativeArray<T>;
```
Comment (wagenet) [inline packages/@ember/-internals/runtime/lib/mixins/array.ts:282]: Previous types didn't allow for the optional value here.

File: packages/@ember/-internals/runtime/lib/mixins/array.ts [hunk 11]
```diff
@@ -209,7 +251,68 @@ function mapBy(key) {
...
+  rejectBy(key: string, value?: unknown): NativeArray<T>;
```
Comment (wagenet) [inline packages/@ember/-internals/runtime/lib/mixins/array.ts:283]: Previous types didn't allow for the optional value here.

File: packages/@ember/-internals/runtime/lib/mixins/array.ts [hunk 12]
```diff
@@ -1344,8 +1476,25 @@ const ArrayMixin = Mixin.create(Enumerable, {
   @uses MutableEnumerable
   @public
 */
-
-const MutableArray = Mixin.create(ArrayMixin, MutableEnumerable, {
+interface MutableArray<T> extends EmberArray<T>, MutableEnumerable {
+  replace(idx: number, amt: number, objects?: readonly T[]): void;
```
Comment (wagenet) [inline packages/@ember/-internals/runtime/lib/mixins/array.ts:1480]: Previous types didn't mark the inbound array as readonly.

File: packages/@ember/-internals/runtime/lib/mixins/array.ts [hunk 13]
```diff
@@ -1775,41 +1929,45 @@ let NativeArray = Mixin.create(MutableArray, Observable, {
...
+  // SAFETY: It's safe to read unknown properties from an object
+  if ((Array.prototype as any)[methodName]) {
     ignore.push(methodName);
   }
 });
 
 NativeArray = NativeArray.without(...ignore);
 
-let A;
+let A: <T>(arr?: Array<T>) => NativeArray<T>;
```
Comment (wagenet) [inline packages/@ember/-internals/runtime/lib/mixins/array.ts:1940]: Previous types were too permissive here for the value of `arr`.

File: packages/@ember/-internals/runtime/lib/mixins/array.ts [hunk 14]
```diff
@@ -1775,41 +1929,45 @@ let NativeArray = Mixin.create(MutableArray, Observable, {
...
-  A = function (arr) {
+  A = function <T>(this: unknown, arr?: Array<T>) {
     assert(
       'You cannot create an Ember Array with `new A()`, please update to calling A as a function: `A()`',
       !(this instanceof A)
     );
 
-    if (!arr) {
-      arr = [];
+    if (isEmberArray(arr)) {
```
Comment (wagenet) [inline packages/@ember/-internals/runtime/lib/mixins/array.ts:1961]: Ideally we'd be more consistent in our use of this method vs EmberArray.detect.

File: packages/@ember/-internals/runtime/lib/system/array_proxy.ts (hunk 8 above)
Comment (chriskrycho) [reply packages/@ember/-internals/runtime/lib/system/array_proxy.ts:214]: Love that we're fixing docs along the way as part of this. :D

File: packages/@ember/-internals/runtime/lib/system/array_proxy.ts [hunk 15]
```diff
@@ -306,8 +305,14 @@ class ArrayProxy<T, C extends EmberArray<T> = EmberArray<T>>
       // @ts-expect-error This check is still good for ensuring correctness
       assert("Can't set ArrayProxy's content to itself", arrangedContent !== this);
       assert(
-        `ArrayProxy expects an Array or ArrayProxy, but you passed ${typeof arrangedContent}`,
-        isArray(arrangedContent) || (arrangedContent as any).isDestroyed
+        `ArrayProxy expects a native Array, EmberArray, or ArrayProxy, but you passed ${typeof arrangedContent}`,
+        (function (arr: unknown): arr is EmberArray<unknown> {
+          return Array.isArray(arr) || EmberArray.detect(arr);
+        })(arrangedContent)
```
Comment (chriskrycho) [inline packages/@ember/-internals/runtime/lib/system/array_proxy.ts:311]: Ha! I independently came up with the type-narrowing-in-`assert`-IIFE around the same time.

File: packages/@ember/object/lib/computed/reduce_computed_macros.ts [hunk 16]
```diff
@@ -966,7 +966,7 @@ export function setDiff(setAProperty: string, setBProperty: string) {
       return emberA();
     }
     if (!isNativeOrEmberArray(setB)) {
-      return emberA(setA);
+      return setA;
```
Comment (chriskrycho) [inline packages/@ember/object/lib/computed/reduce_computed_macros.ts:969]: Huh. Are you sure?

File: packages/@ember/-internals/runtime/lib/mixins/array.ts [hunk 17]
```diff
@@ -22,17 +22,26 @@ import { ENV } from '@ember/-internals/environment';
...
 import { typeOf } from '../type-of';
+import type { AnyFn } from '@ember/-internals/utils/types';
+import type { ComputedPropertyCallback } from '@ember/-internals/metal/lib/computed';
 
-const EMPTY_ARRAY = Object.freeze([]);
+type Value<T, K extends string> = K extends keyof T ? T[K] : unknown;
 
-const identityFunction = (item) => item;
+// SAFETY: Nothing unsafe here, just TS can't infer the specificity we can.
+const EMPTY_ARRAY = Object.freeze([]) as readonly [];
```
Comment (chriskrycho) [inline packages/@ember/-internals/runtime/lib/mixins/array.ts:31]: This is a great place to use `as const`, which doesn't require the `SAFETY` annotation (`as const` is never an unsafe cast) and gives *exactly* the desired type ([playground][p]).

```suggestion
const EMPTY_ARRAY = Object.freeze([] as const);
```

[p]: https://www.typescriptlang.org/play?#code/MYewdgzgLgBAogWQAoBUCaB9AggJR1tGAXhgHkAjAKwFNgoA6AMwCdrqAvagCgG0BdGAEMIMUJCgBKANxA

File: packages/@ember/-internals/runtime/lib/mixins/array.ts [hunk 18]
```diff
@@ -22,17 +22,26 @@ import { ENV } from '@ember/-internals/environment';
...
+
+export function uniqBy<T>(
+  array: T[] | EmberArray<T>,
+  keyOrFunc: string | ((item: T) => unknown) = identityFunction
+): T[] | EmberArray<T> {
   assert(`first argument passed to \`uniqBy\` should be array`, isArray(array));
 
-  let ret = A();
+  // SAFETY: This created an empty array so casting it to our known type is safe.
+  let ret = A() as NativeArray<T>;
```
Comment (chriskrycho) [inline packages/@ember/-internals/runtime/lib/mixins/array.ts:42]: Huh. Is `A()` not generic?

File: packages/@ember/-internals/runtime/lib/mixins/array.ts (hunk 13 above)
Comment (chriskrycho) [reply packages/@ember/-internals/runtime/lib/mixins/array.ts:1940]: It *is* generic! At the other site, can you just `= A<T>()`? (What a mess of characters that is, though 😂)

File: packages/@ember/-internals/runtime/lib/mixins/array.ts (hunk 14 above)
Comment (chriskrycho) [reply packages/@ember/-internals/runtime/lib/mixins/array.ts:1961]: Ideally we will soon burn all of this out of existence. 😏

**Review by chriskrycho — COMMENTED** [review]
Finished a review pass!

File: packages/@ember/-internals/runtime/lib/mixins/array.ts (hunk 17 above)
Comment (wagenet) [reply packages/@ember/-internals/runtime/lib/mixins/array.ts:31]: Great, thanks!

File: packages/@ember/-internals/runtime/lib/mixins/array.ts (hunk 18 above)
Comment (wagenet) [reply packages/@ember/-internals/runtime/lib/mixins/array.ts:42]: I refactored the types for `A` later and missed this.

File: packages/@ember/-internals/runtime/lib/mixins/array.ts (hunk 13 above)
Comment (wagenet) [reply packages/@ember/-internals/runtime/lib/mixins/array.ts:1940]: Yeah, I fixed the other spot.

File: packages/@ember/object/lib/computed/reduce_computed_macros.ts (hunk 16 above)
Comment (wagenet) [reply packages/@ember/object/lib/computed/reduce_computed_macros.ts:969]: @chriskrycho and I talked elsewhere. This is the summary in case anyone else stumbles across this:

tl;dr emberA(setA) is a type error
I don't think you can correctly convert a non-js array EmberArray to a NativeArray
prior to my change, setDiff claims to return unknown[] | NativeArray<unknown>
after removing the emberA it becomes unknown[] | EmberArray<unknown> which I think is actually correct enough
if we really wanted a NativeArray we could do emberA(setA.toArray()) or equivalent
docs just say "computes a new array

File: packages/@ember/-internals/runtime/lib/mixins/array.ts (hunk 14 above)
Comment (runspired) [reply packages/@ember/-internals/runtime/lib/mixins/array.ts:1959]: Sadly this breaks EmberData's migration path :( 

Data Issue: https://github.com/emberjs/data/issues/8202
PR to restore: https://github.com/emberjs/ember.js/pull/20219
